README.md
# OptionParser_v2
A C++20 command line parser library. 

## Release checklist
1. Update project version in `CMakeLists.txt`
2. Update `CHANGELOG.md`
3. Update `README.md` if needed
4. Create tag and push to github
5. Create release on github

## How to format code
Use clang-format with the provided .clang-format file.

Example from repo root:
```bash
clang-format -i src/*pp test/*pp
```

# Legacy OptionParser
A C++17 command line parser library.
The goal is to extract simple structure out of strings in a similar fashion to how *nix command line programs use flags to specify options.
To use the library, specify how you want the structure of the string to look in the template parameters, and a parser will be generated.
					
Thanks to meta-template-programming, OptionParser will generate a parser based on the parameters at compile time, which allows the parser to be used at runtime with no startup cost, and minimal parse-time cost (we can apply compile time optimizations to the spesific parser).
A spesific implemtation optimization is the use of C++17 std::string_view, to have no heap allocations while parsing (the maximum amount of options is known at compile time). There are no stored copies of parsed results, only pointers to the original string. The rationale for this optimization is to reduce size and improve cache reuse and locality.
The library is statically typed, which means that you will get a compile time error if you try to access a result member(option) that is not part of the parser generated.

## Requirements
C++20 or later compatible compiler.

## Examples
How a mini version of the gcc command (man gcc) would be implemented in OptionParser. (In some of the gcc options, the option and arguments are not separated. OptionParser is always using spaces.)
(It is not possible to have multiple instances the same option type in OptionParser, so the "-D" option can not be properly implemented in OptionParser at this time, instead the option is implemented as a ListType and then overloaded with a WordType.
```cpp
using namespace OptionParser;
Parser parser(
	Option<WordType>("gcc"),
	Option<>("-c"),
	Option<>("-S"),
	Option<>("-E"),
	Option<WordType>("-std"),
	Option<>("-g"),
	Option<>("-pg"),
	Option<NumberType<int>>("-O"),
	Option<ListType>("-W"),
	Option<WordType>("-W"),
	Option<>("-pedantic"),
	Option<ListType>("-I"),
	Option<WordType>("-I"),
	Option<ListType>("-L"),
	Option<WordType>("-L"),
	Option<ListType>("-D"),
	Option<WordType>("-D"),
	Option<WordType>("-U"),
	Option<ListType>("-f"),
	Option<WordType>("-f"),
	Option<ListType>("-m"),
	Option<WordType>("-m"),
	Option<WordType>("-o")
);

//try to change this string
std::string input_str = "gcc main.cpp -o main -std c++17 -O 3 -D [EXAMPLE_MACRO1=0x1010, EXAMPLE_MACRO2=TEST]";

auto set = parser.parse(input_str);

if (auto gcc_result = set.find<0>()) {
	std::cout << "gcc argument: " << gcc_result->get<0>() << "\n";

	if (set.find<1>()) {
		std::cout << "-c option\n";
	}
	else if (set.find<2>()) {
		std::cout << "-S option\n";
	}
	else if (set.find<3>()) {
		std::cout << "-E option\n";
	}

	if (auto result_std = set.find<4>()) {
		std::cout << "-std option: " << result_std->get<0>() << "\n";
	}

	if (set.find<5>()) {
		std::cout << "-g option\n";
	}

	if (set.find<6>()) {
		std::cout << "-pg option\n";
	}

	if (auto result_O = set.find<7>()) {
		std::cout << "-O option: " << result_O->get<0>() << "\n";
	}

	if (auto result_W = set.find<8>()) {
		std::cout << "-W option:\n";
		for (auto m : result_W->get<0>()) {
			std::cout << "\t" << m << "\n";
		}
	}
	else if (auto result_W = set.find<9>()) {
		std::cout << "-W option: " << result_W->get<0>() << "\n";
	}

	if (auto result_pedantic = set.find<10>()) {
		std::cout << "-pedantic option\n";
	}

	if (auto result_I = set.find<11>()) {
		std::cout << "-I option:\n";
		for (auto m : result_I->get<0>()) {
			std::cout << "\t" << m << "\n";
		}
	}
	else if (auto result_I = set.find<12>()) {
		std::cout << "-I option: " << result_I->get<0>() << "\n";
	}

	if (auto result_L = set.find<13>()) {
		std::cout << "-L option:\n";
		for (auto m : result_L->get<0>()) {
			std::cout << "\t" << m << "\n";
		}
	}
	else if (auto result_L = set.find<14>()) {
		std::cout << "-L option: " << result_L->get<0>() << "\n";
	}

	if (auto result_D = set.find<15>()) {
		std::cout << "-D option:\n";
		for (auto m : result_D->get<0>()) {
			std::cout << "\t" << m << "\n";
		}
	}
	else if (auto result_D = set.find<16>()) {
		std::cout << "-D option: " << result_D->get<0>() << "\n";
	}

	if (auto result_U = set.find<17>()) {
		std::cout << "-U option: " << result_U->get<0>() << "\n";
	}

	if (auto result_f = set.find<18>()) {
		std::cout << "-f option:\n";
		for (auto m : result_f->get<0>()) {
			std::cout << "\t" << m << "\n";
		}
	}
	else if (auto result_f = set.find<19>()) {
		std::cout << "-f option: " << result_f->get<0>() << "\n";
	}

	if (auto result_m = set.find<20>()) {
		std::cout << "-m option:\n";
		for (auto m : result_m->get<0>()) {
			std::cout << "\t" << m << "\n";
		}
	}
	else if (auto result_m = set.find<21>()) {
		std::cout << "-m option: " << result_m->get<0>() << "\n";
	}

	if (auto result_o = set.find<22>()) {
		std::cout << "-o option: " << result_o->get<0>() << "\n";
	}

}
```
```
Output:
gcc argument: main.cpp
-std option: c++17
-O option: 3
-D option:
        EXAMPLE_MACRO1=0x1010
        EXAMPLE_MACRO2=TEST
-o option: main
```

Only capture flag if the input string fulfills the option parameter requirements. 
```cpp
using namespace OptionParser;
Parser parser(Option<StringType, NumberType<int>>("-flag1"));
std::string str1 = "-flag1 -flag1 \"this is a string\" 123 -flag1";
std::string str2 = "-flag1 -flag1 -flag1 -flag1 -flag1";
auto set1 = parser.parse(str1);
if (auto result = set1.find<0>()) {
	std::cout << result->get<0>() << " " << result->get<1>() << "\n";
}
else {
	std::cout << "Did not find flag1 in str1\n";
}
auto set2 = parser.parse(str2);
if (auto result = set2.find<0>()) {
	std::cout << result->get<0>() << " " << result->get<1>() << "\n";
}
else {
	std::cout << "Did not find flag1 in str2\n";
}
```
	Output:
	this is a string 123
	Did not find flag1 in str2

Overload flags.
```cpp
using namespace OptionParser;
Parser parser(Option<NumberType<int>>("-flag1"), Option<WordType>("-flag1"));
std::string str1 = "-flag1 12345";
std::string str2 = "-flag1 word123";
auto set1 = parser.parse(str1);
if (auto result = set1.find<0>()) {
	std::cout << "NumberType: " << result->get<0>() << "\n";
}
else if (auto result = set1.find<1>()) {
	std::cout << "WordType: " << result->get<0>() << "\n";
}
else {
	std::cout << "Not found\n";
}
auto set2 = parser.parse(str2);
if (auto result = set2.find<0>()) {
	std::cout << "NumberType: " << result->get<0>() << "\n";
}
else if (auto result = set2.find<1>()) {
	std::cout << "WordType: " << result->get<0>() << "\n";
}
else {
	std::cout << "Not found\n";
}
```
	Output:
	NumberType: 12345
	WordType: word123

	

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [v0.5.0] - 2023-8-19
### Added
- Added reference version of multi functions in `optionparser_v2`. (https://github.com/sivertac/OptionParser/pull/13)

## [v0.4.0] - 2023-8-19
### Added
- Added optional Required Components, parser will validate if required components are present. (https://github.com/sivertac/OptionParser/pull/12)
    - `isComplete()` member function in `ParseContext` will return `false` if required components are missing for the current `ParseResult`.
    - `makeRequiredParameter(...)`, `makeRequiredFlag(...)`, `makeRequiredCommand(...)` functions to make required components.
- Added automatic help text generators for `optionparser_v2` Components. (https://github.com/sivertac/OptionParser/pull/12)
    - `generateUsageString` will generate a usage string for the parser.
    - `generateHelpString` will generate a help string for the parser.

## [v0.3.1] - 2023-8-17
### Fixed
- Fixed wrong suggestions when all parameters of a component has been consumed if the parameters has a custom suggestions function, now parameters will only be suggested if there are space for more parameters in the result. (https://github.com/sivertac/OptionParser/pull/11)

## [v0.3.0] - 2023-8-17
### Added
- Added `ParseContext` to `optionparser_v2`. (https://github.com/sivertac/OptionParser/pull/8)
    - `ParseContext` is used to store the state of the parser, allowing tokens to be fed to the parser one at a time.

### Changed
- Changed stored value type in `ParseResult` from `std::string_view` to `std::string` in `optionparser_v2`. (https://github.com/sivertac/OptionParser/pull/9)
    - This is to avoid dangling references to the original input string.

## [v0.2.1]
### Fixed
- Set correct project version in `CMakeLists.txt`.

## [v0.2.0] - 2023-8-15 [YANKED]
### Added
- Added `nextTokenSuggestionsMulti` function to `optionparser_v2`. (https://github.com/sivertac/OptionParser/pull/2)
- Added `CHANGELOG.md` file.

### Changed
- Changed name of `PositionalIdentifier` to `Command` in `optionparser_v2`. (https://github.com/sivertac/OptionParser/pull/3)


## [v0.1.0] - 2023-8-13
### Added
- Added `optionparser_v2`
    - Runtime capable parser.
    - Supports parse suggestions that can be used for tab completion.
    - Everything is new, nothing is shared with legacy `OptionParser`.
### Changed
- Tests now use `google-test` instead of `doctest`.


## [v0.0.1] - 2020-10-31
- Legacy `OptionParser`.


test/test1.cpp

#include <doctest/doctest.h>

#include <unistd.h>


SCENARIO("sleep a second") {

  GIVEN("a one second sleep duration") {
    unsigned int duration =1 ;

    WHEN("call sleep with this duration") {
      int ret = sleep(duration) ;

      THEN("it's expected nobody interupted our rest") {
        CHECK(ret == 0);
      }
    }
  }
} 


test/test2.cpp

#include <OptionParser.hpp>
#include <string>
#include <unistd.h>

#include <gtest/gtest.h>

TEST(OptionParser, test1) {
    using namespace OptionParser;
    Parser parser(Option<NumberType<int>>("number"));

    std::string input_string = "number 123";

    auto set = parser.parse(input_string);
    if (auto res = set.find<0>()) {
        EXPECT_EQ(res->get<0>(), 123);
    }
}


test/test_main.cpp
#define DOCTEST_CONFIG_IMPLEMENT
#include <doctest/doctest.h>

int main(int argc, char** argv) {
    doctest::Context context;
    context.applyCommandLine(argc, argv);
    return context.run(); 
}




src/main.cpp
//main.cpp

//C++
#include <iostream>
#include <string>

//Local
#include "../include/OptionParser.hpp"

#if 1

int main(int argc, char** argv)
{
	using namespace OptionParser;
	Parser parser(
		Option<WordType>("gcc"),
		Option<>("-c"),
		Option<>("-S"),
		Option<>("-E"),
		Option<WordType>("-std"),
		Option<>("-g"),
		Option<>("-pg"),
		Option<NumberType<int>>("-O"),
		Option<ListType>("-W"),
		Option<WordType>("-W"),
		Option<>("-pedantic"),
		Option<ListType>("-I"),
		Option<WordType>("-I"),
		Option<ListType>("-L"),
		Option<WordType>("-L"),
		Option<ListType>("-D"),
		Option<WordType>("-D"),
		Option<WordType>("-U"),
		Option<ListType>("-f"),
		Option<WordType>("-f"),
		Option<ListType>("-m"),
		Option<WordType>("-m"),
		Option<WordType>("-o")
	);

	std::string input_str = "gcc main.cpp -o main -std c++17 -O 3 -D [EXAMPLE_MACRO1=0x1010, EXAMPLE_MACRO2=TEST]";

	auto set = parser.parse(input_str);

	if (auto gcc_result = set.find<0>()) {
		std::cout << "gcc argument: " << gcc_result->get<0>() << "\n";

		if (set.find<1>()) {
			std::cout << "-c option\n";
		}
		else if (set.find<2>()) {
			std::cout << "-S option\n";
		}
		else if (set.find<3>()) {
			std::cout << "-E option\n";
		}

		if (auto result_std = set.find<4>()) {
			std::cout << "-std option: " << result_std->get<0>() << "\n";
		}

		if (set.find<5>()) {
			std::cout << "-g option\n";
		}

		if (set.find<6>()) {
			std::cout << "-pg option\n";
		}

		if (auto result_O = set.find<7>()) {
			std::cout << "-O option: " << result_O->get<0>() << "\n";
		}

		if (auto result_W = set.find<8>()) {
			std::cout << "-W option:\n";
			for (auto m : result_W->get<0>()) {
				std::cout << "\t" << m << "\n";
			}
		}
		else if (auto result_W = set.find<9>()) {
			std::cout << "-W option: " << result_W->get<0>() << "\n";
		}

		if (auto result_pedantic = set.find<10>()) {
			std::cout << "-pedantic option\n";
		}

		if (auto result_I = set.find<11>()) {
			std::cout << "-I option:\n";
			for (auto m : result_I->get<0>()) {
				std::cout << "\t" << m << "\n";
			}
		}
		else if (auto result_I = set.find<12>()) {
			std::cout << "-I option: " << result_I->get<0>() << "\n";
		}

		if (auto result_L = set.find<13>()) {
			std::cout << "-L option:\n";
			for (auto m : result_L->get<0>()) {
				std::cout << "\t" << m << "\n";
			}
		}
		else if (auto result_L = set.find<14>()) {
			std::cout << "-L option: " << result_L->get<0>() << "\n";
		}

		if (auto result_D = set.find<15>()) {
			std::cout << "-D option:\n";
			for (auto m : result_D->get<0>()) {
				std::cout << "\t" << m << "\n";
			}
		}
		else if (auto result_D = set.find<16>()) {
			std::cout << "-D option: " << result_D->get<0>() << "\n";
		}

		if (auto result_U = set.find<17>()) {
			std::cout << "-U option: " << result_U->get<0>() << "\n";
		}

		if (auto result_f = set.find<18>()) {
			std::cout << "-f option:\n";
			for (auto m : result_f->get<0>()) {
				std::cout << "\t" << m << "\n";
			}
		}
		else if (auto result_f = set.find<19>()) {
			std::cout << "-f option: " << result_f->get<0>() << "\n";
		}

		if (auto result_m = set.find<20>()) {
			std::cout << "-m option:\n";
			for (auto m : result_m->get<0>()) {
				std::cout << "\t" << m << "\n";
			}
		}
		else if (auto result_m = set.find<21>()) {
			std::cout << "-m option: " << result_m->get<0>() << "\n";
		}

		if (auto result_o = set.find<22>()) {
			std::cout << "-o option: " << result_o->get<0>() << "\n";
		}

	}



	return 0;
}

#endif

#if 0
/*
Overload flags.
*/
int main(int argc, char** argv)
{
	using namespace OptionParser;
	Parser parser(Option<NumberType<int>>("-flag1"), Option<WordType>("-flag1"));

	std::string str1 = "-flag1 12345";
	std::string str2 = "-flag1 word123";

	auto set1 = parser.parse(str1);

	if (auto result = set1.find<0>()) {
		std::cout << "NumberType: " << result->get<0>() << "\n";
	}
	else if (auto result = set1.find<1>()) {
		std::cout << "WordType: " << result->get<0>() << "\n";
	}
	else {
		std::cout << "Not found\n";
	}

	auto set2 = parser.parse(str2);

	if (auto result = set2.find<0>()) {
		std::cout << "NumberType: " << result->get<0>() << "\n";
	}
	else if (auto result = set2.find<1>()) {
		std::cout << "WordType: " << result->get<0>() << "\n";
	}
	else {
		std::cout << "Not found\n";
	}

	return 0;
}

#endif

#if 0
/*
Only capture flag if the input string fulfills the option parameter requirements.
*/
int main(int argc, char** argv)
{
	using namespace OptionParser;
	Parser parser(Option<StringType, NumberType<int>>("-flag1"));

	std::string str1 = "-flag1 -flag1 \"this is a string\" 123 -flag1";
	std::string str2 = "-flag1 -flag1 -flag1 -flag1 -flag1";

	auto set1 = parser.parse(str1);

	if (auto result = set1.find<0>()) {
		std::cout << result->get<0>() << " " << result->get<1>() << "\n";
	}
	else {
		std::cout << "Did not find flag1 in str1\n";
	}

	auto set2 = parser.parse(str2);

	if (auto result = set2.find<0>()) {
		std::cout << result->get<0>() << " " << result->get<1>() << "\n";
	}
	else {
		std::cout << "Did not find flag1 in str2\n";
	}

	return 0;
}

#endif

#if 0

template <typename Iterator>
void print(Iterator begin, Iterator end)
{
	while (begin != end) {
		if (begin < end - 1) {
			std::cout << *begin << ", ";
		}
		else {
			std::cout << *begin << "\n";
		}

		std::advance(begin, 1);
	}
}

int main(int argc, char** argv)
{

	std::cout << std::boolalpha;

	using namespace OptionParser;

	Parser parser(Option<NumberType<int>>("-int"), Option<NumberType<double>>("-double"));

	//std::string_view str1("-int 123.123213 -double -123");
	std::string_view str1("-double -123");
	//std::string_view str1(" a -1 \" -0 sgjerogjer \" -2 -0 b    ");

	auto set = parser.parse(str1);

	if (auto r = set.find<0>()) {
		std::cout << "found 0\n";
		auto l = r->get<0>();
		std::cout << l << "\n";
	}
	if (auto r = set.find<1>()) {
		std::cout << "found 1\n";
		auto l = r->get<0>();
		std::cout << l << "\n";
	}

	//if (auto r = set.find<2>()) {
	//	std::cout << "found 2\n";
	//}
	//if (auto r = set.find<3>()) {
	//	std::cout << "found 3\n";
	//}

	return 0;
}
#endif

#if 0

int main(int argc, char** argv)
{
	OptionParser parser(Option("-0"), Option("-1", 1), Option("-2"));

	auto set = parser.parse("-1 -2 nam -2 -2 -1 -3");

	if (auto r = set.find(parser.get<0>())) {
		std::cout << "found 0" << "\n";
	}
	if (auto r = set.find(parser.get<1>())) {
		std::cout << "found 1" << "\n";
	}
	if (auto r = set.find(parser.get<2>())) {
		std::cout << "found 2" << "\n";
	}




	return 0;
}
#endif

src/OptionParser.cpp

#include "OptionParser.hpp"

namespace OptionParser {

std::vector<std::string_view> extractWords(std::string_view str_view,
                                           const std::string_view &delim) {
    std::vector<std::string_view> word_vec;
    while (!str_view.empty()) {
        std::size_t delim_pos = str_view.find_first_of(delim);
        std::size_t word_pos = str_view.find_first_not_of(delim);
        if (delim_pos == str_view.npos) {
            word_vec.push_back(str_view);
            break;
        } else if (word_pos == str_view.npos) {
            break;
        } else if (delim_pos > word_pos) {
            word_vec.push_back(str_view.substr(0, delim_pos));
            str_view.remove_prefix(delim_pos);
        } else {
            str_view.remove_prefix(word_pos);
        }
    }
    return word_vec;
}

std::pair<std::string_view, std::size_t>
removePrefixDelim(std::string_view str_view, const std::string_view &delim) {
    auto delim_pos = str_view.find_first_of(delim);
    if (delim_pos == str_view.npos)
        return std::make_pair(str_view, 0);
    auto word_pos = str_view.find_first_not_of(delim);
    if (word_pos == str_view.npos) {
        std::size_t s = str_view.size();
        str_view.remove_prefix(str_view.size());
        return std::make_pair(str_view, s);
    }
    if (word_pos < delim_pos) {
        return std::make_pair(str_view, 0);
    } else {
        str_view.remove_prefix(word_pos);
        return std::make_pair(str_view, word_pos);
    }
}

std::pair<std::string_view, std::size_t>
extractFirstWord(std::string_view str_view, const std::string_view &delim) {
    auto pair = removePrefixDelim(str_view, delim);
    auto delim_pos = pair.first.find_first_of(delim);
    if (delim_pos == pair.first.npos)
        return std::make_pair(pair.first, pair.second + pair.first.size());
    auto word_pos = pair.first.find_first_not_of(delim);
    if (word_pos == pair.first.npos) {
        str_view.remove_prefix(pair.first.size());
        return std::make_pair(pair.first, pair.second);
    }
    return std::make_pair(pair.first.substr(0, delim_pos),
                          pair.second + delim_pos);
}

bool hasPrefixDelim(std::string_view str_view, const std::string_view &delim) {
    auto delim_pos = str_view.find_first_of(delim);
    auto word_pos = str_view.find_first_not_of(delim);
    if (delim_pos == str_view.npos && word_pos == str_view.npos) {
        return false;
    } else if (delim_pos == str_view.npos) {
        return false;
    } else if (word_pos == str_view.npos) {
        return true;
    } else if (delim_pos > word_pos) {
        return false;
    } else /* if (delim_pos < word_pos) */ {
        return true;
    }
}

std::pair<std::string_view, std::size_t>
removeFirstWord(std::string_view str_view, const std::string_view &delim) {
    std::size_t s = 0;
    auto pair = removePrefixDelim(str_view, delim);
    s += pair.second;
    auto first_word = extractFirstWord(pair.first, delim);
    pair.first.remove_prefix(first_word.first.size());
    s += first_word.first.size();
    pair = removePrefixDelim(pair.first, delim);
    s += pair.second;
    return std::make_pair(pair.first, s);
}

std::optional<std::string_view>
extractFirstWordDestructive(std::string_view &str_view,
                            const std::string_view &delim) {
    auto pair1 = extractFirstWord(str_view, delim);
    if (pair1.first.empty()) {
        return std::nullopt;
    } else {
        auto pair2 = removeFirstWord(str_view, delim);
        str_view = pair2.first;
        return pair1.first;
    }
}

std::size_t searchEndOfList(std::string_view str_view, char begin, char end) {
    assert(begin != end);

    if (str_view.empty())
        return str_view.npos;

    str_view.remove_prefix(1); // remove first begin
    std::size_t count = 1;

    std::size_t total_pos = 0;

    while (count > 0) {
        auto end_pos = str_view.find(end);
        if (end_pos == str_view.npos)
            return str_view.npos;
        auto begin_pos = str_view.find(begin);
        if (begin_pos < end_pos) {
            str_view.remove_prefix(begin_pos + 1);
            total_pos += begin_pos + 1;
            ++count;
        } else {
            str_view.remove_prefix(end_pos + 1);
            total_pos += end_pos + 1;
            --count;
        }
    }
    return total_pos;
}

std::pair<std::string_view, std::size_t>
extractFirstList(std::string_view str_view, char begin, char end,
                 const std::string_view &delim) {
    assert(begin != end);
    auto pair = removePrefixDelim(str_view, delim);
    auto end_pos = searchEndOfList(pair.first, begin, end);
    if (end_pos == pair.first.npos) {
        return std::make_pair(pair.first.substr(0, 0), pair.second);
    } else {
        return std::make_pair(pair.first.substr(0, end_pos + 1),
                              pair.second + end_pos + 1);
    }
}

std::pair<std::string_view, std::size_t>
extractFirstString(std::string_view str_view, const std::string_view &delim) {
    auto pair = removePrefixDelim(str_view, delim);

    if (pair.first.empty() || pair.first.front() != '"') {
        return std::make_pair(pair.first.substr(0, 0), pair.second);
    }

    pair.first.remove_prefix(1);
    auto end = pair.first.find_first_of('"');
    if (end == pair.first.npos) {
        return std::make_pair(pair.first.substr(0, 0), pair.second + 1);
    }
    return std::make_pair(pair.first.substr(0, end), pair.second + 1 + end + 1);
}

} // namespace OptionParser


examples/optionparser_v2_help_text.cpp

#include <OptionParser_v2.hpp>
#include <iostream>

int main(int argc, char **argv) {

    using namespace optionparser_v2;

    Component root_component = makeCommand(
        "root", "root command",
        {makeFlag("--flag", "", "flag description"),
         makeParameter("parameter", "parameter description"),
         makeFlag("--flag2", "", "flag2 description",
                  {makeRequiredParameter("flag2_parameter",
                                         "flag2 parameter description")}),
         makeRequiredCommand(
             "subcommand", "subcommand description",
             {makeFlag("--subcommand_flag", "-s",
                       "subcommand flag description"),
              makeParameter("subcommand_parameter",
                            "subcommand parameter description")})});

    std::cout << generateUsageString(root_component) << std::endl;

    std::cout << generateHelpString(root_component) << std::endl;

    return 0;
}


examples/gcc_command_line_example.cpp
// main.cpp

// C++
#include <iostream>
#include <string>

// Local
#include "../include/OptionParser.hpp"

#if 1

int main(int argc, char **argv) {
    using namespace OptionParser;
    Parser parser(
        Option<WordType>("gcc"), Option<>("-c"), Option<>("-S"), Option<>("-E"),
        Option<WordType>("-std"), Option<>("-g"), Option<>("-pg"),
        Option<NumberType<int>>("-O"), Option<ListType>("-W"),
        Option<WordType>("-W"), Option<>("-pedantic"), Option<ListType>("-I"),
        Option<WordType>("-I"), Option<ListType>("-L"), Option<WordType>("-L"),
        Option<ListType>("-D"), Option<WordType>("-D"), Option<WordType>("-U"),
        Option<ListType>("-f"), Option<WordType>("-f"), Option<ListType>("-m"),
        Option<WordType>("-m"), Option<WordType>("-o"));

    std::string input_str = "gcc main.cpp -o main -std c++17 -O 3 -D "
                            "[EXAMPLE_MACRO1=0x1010, EXAMPLE_MACRO2=TEST]";

    auto set = parser.parse(input_str);

    if (auto gcc_result = set.find<0>()) {
        std::cout << "gcc argument: " << gcc_result->get<0>() << "\n";

        if (set.find<1>()) {
            std::cout << "-c option\n";
        } else if (set.find<2>()) {
            std::cout << "-S option\n";
        } else if (set.find<3>()) {
            std::cout << "-E option\n";
        }

        if (auto result_std = set.find<4>()) {
            std::cout << "-std option: " << result_std->get<0>() << "\n";
        }

        if (set.find<5>()) {
            std::cout << "-g option\n";
        }

        if (set.find<6>()) {
            std::cout << "-pg option\n";
        }

        if (auto result_O = set.find<7>()) {
            std::cout << "-O option: " << result_O->get<0>() << "\n";
        }

        if (auto result_W = set.find<8>()) {
            std::cout << "-W option:\n";
            for (auto m : result_W->get<0>()) {
                std::cout << "\t" << m << "\n";
            }
        } else if (auto result_W = set.find<9>()) {
            std::cout << "-W option: " << result_W->get<0>() << "\n";
        }

        if (auto result_pedantic = set.find<10>()) {
            std::cout << "-pedantic option\n";
        }

        if (auto result_I = set.find<11>()) {
            std::cout << "-I option:\n";
            for (auto m : result_I->get<0>()) {
                std::cout << "\t" << m << "\n";
            }
        } else if (auto result_I = set.find<12>()) {
            std::cout << "-I option: " << result_I->get<0>() << "\n";
        }

        if (auto result_L = set.find<13>()) {
            std::cout << "-L option:\n";
            for (auto m : result_L->get<0>()) {
                std::cout << "\t" << m << "\n";
            }
        } else if (auto result_L = set.find<14>()) {
            std::cout << "-L option: " << result_L->get<0>() << "\n";
        }

        if (auto result_D = set.find<15>()) {
            std::cout << "-D option:\n";
            for (auto m : result_D->get<0>()) {
                std::cout << "\t" << m << "\n";
            }
        } else if (auto result_D = set.find<16>()) {
            std::cout << "-D option: " << result_D->get<0>() << "\n";
        }

        if (auto result_U = set.find<17>()) {
            std::cout << "-U option: " << result_U->get<0>() << "\n";
        }

        if (auto result_f = set.find<18>()) {
            std::cout << "-f option:\n";
            for (auto m : result_f->get<0>()) {
                std::cout << "\t" << m << "\n";
            }
        } else if (auto result_f = set.find<19>()) {
            std::cout << "-f option: " << result_f->get<0>() << "\n";
        }

        if (auto result_m = set.find<20>()) {
            std::cout << "-m option:\n";
            for (auto m : result_m->get<0>()) {
                std::cout << "\t" << m << "\n";
            }
        } else if (auto result_m = set.find<21>()) {
            std::cout << "-m option: " << result_m->get<0>() << "\n";
        }

        if (auto result_o = set.find<22>()) {
            std::cout << "-o option: " << result_o->get<0>() << "\n";
        }
    }

    return 0;
}

#endif

#if 0
/*
Overload flags.
*/
int main(int argc, char** argv)
{
	using namespace OptionParser;
	Parser parser(Option<NumberType<int>>("-flag1"), Option<WordType>("-flag1"));

	std::string str1 = "-flag1 12345";
	std::string str2 = "-flag1 word123";

	auto set1 = parser.parse(str1);

	if (auto result = set1.find<0>()) {
		std::cout << "NumberType: " << result->get<0>() << "\n";
	}
	else if (auto result = set1.find<1>()) {
		std::cout << "WordType: " << result->get<0>() << "\n";
	}
	else {
		std::cout << "Not found\n";
	}

	auto set2 = parser.parse(str2);

	if (auto result = set2.find<0>()) {
		std::cout << "NumberType: " << result->get<0>() << "\n";
	}
	else if (auto result = set2.find<1>()) {
		std::cout << "WordType: " << result->get<0>() << "\n";
	}
	else {
		std::cout << "Not found\n";
	}

	return 0;
}

#endif

#if 0
/*
Only capture flag if the input string fulfills the option parameter requirements.
*/
int main(int argc, char** argv)
{
	using namespace OptionParser;
	Parser parser(Option<StringType, NumberType<int>>("-flag1"));

	std::string str1 = "-flag1 -flag1 \"this is a string\" 123 -flag1";
	std::string str2 = "-flag1 -flag1 -flag1 -flag1 -flag1";

	auto set1 = parser.parse(str1);

	if (auto result = set1.find<0>()) {
		std::cout << result->get<0>() << " " << result->get<1>() << "\n";
	}
	else {
		std::cout << "Did not find flag1 in str1\n";
	}

	auto set2 = parser.parse(str2);

	if (auto result = set2.find<0>()) {
		std::cout << result->get<0>() << " " << result->get<1>() << "\n";
	}
	else {
		std::cout << "Did not find flag1 in str2\n";
	}

	return 0;
}

#endif

#if 0

template <typename Iterator>
void print(Iterator begin, Iterator end)
{
	while (begin != end) {
		if (begin < end - 1) {
			std::cout << *begin << ", ";
		}
		else {
			std::cout << *begin << "\n";
		}

		std::advance(begin, 1);
	}
}

int main(int argc, char** argv)
{

	std::cout << std::boolalpha;

	using namespace OptionParser;

	Parser parser(Option<NumberType<int>>("-int"), Option<NumberType<double>>("-double"));

	//std::string_view str1("-int 123.123213 -double -123");
	std::string_view str1("-double -123");
	//std::string_view str1(" a -1 \" -0 sgjerogjer \" -2 -0 b    ");

	auto set = parser.parse(str1);

	if (auto r = set.find<0>()) {
		std::cout << "found 0\n";
		auto l = r->get<0>();
		std::cout << l << "\n";
	}
	if (auto r = set.find<1>()) {
		std::cout << "found 1\n";
		auto l = r->get<0>();
		std::cout << l << "\n";
	}

	//if (auto r = set.find<2>()) {
	//	std::cout << "found 2\n";
	//}
	//if (auto r = set.find<3>()) {
	//	std::cout << "found 3\n";
	//}

	return 0;
}
#endif

#if 0

int main(int argc, char** argv)
{
	OptionParser parser(Option("-0"), Option("-1", 1), Option("-2"));

	auto set = parser.parse("-1 -2 nam -2 -2 -1 -3");

	if (auto r = set.find(parser.get<0>())) {
		std::cout << "found 0" << "\n";
	}
	if (auto r = set.find(parser.get<1>())) {
		std::cout << "found 1" << "\n";
	}
	if (auto r = set.find(parser.get<2>())) {
		std::cout << "found 2" << "\n";
	}




	return 0;
}
#endif